src/kernel-install/ukify.py

```diff
@@ -0,0 +1,472 @@
+    offset: int = dataclasses.field(default=None, init=False)
+
+    def __post_init__(self):
+        self.offset = round_to_page(pe_executable_size(self.executable))
+
+    def add_section(self, section):
+        assert self.offset
+        assert section.offset is None
+        section.offset = self.offset
+        self.offset += round_to_page(section.size())
+        self.sections += [section]
+
+
+def parse_banks(s):
+    banks = s.split(',')
+    # TODO: do some sanity checking here
+    return banks
+
+
+def check_splash(args):
+    if args.splash is None:
+        return
+
+    # import is delayed, to avoid import when the splash image is not used
+    try:
+        from PIL import Image
+    except ImportError:
+        return
+
+    img = Image.open(args.splash)
+    print(f'Splash image {args.splash.name} is {img.width}×{img.height} pixels')
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name in {'output', 'tools'}:
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        value.open().close()
+
+
+def find_tool(args, name, fallback=None):
+    tool = args.tools / name
+    if tool.exists():
+        return tool
+    return fallback or name
+
+
+def make_uki(args):
+    # kernel payload signing
+
+    sbsign_tool = find_tool(args, 'sbsign')
+    sbsign_invocation = [
+        sbsign_tool,
+        '--key', args.sb_key,
+        '--cert', args.sb_cert,
+    ]
+
+    sign_kernel = args.sign_kernel
+    if sign_kernel is None and args.sb_key:
+        # figure out if we should sign the kernel
+        sbverify_tool = find_tool(args, 'sbverify')
+
+        cmd = [
+            sbverify_tool,
+            '--list',
+            args.linux,
+        ]
+
+        print('+', shell_join(cmd))
+        info = subprocess.check_output(cmd, text=True)
+
+        # sbverify has wonderful API
+        if 'No signature table present' in info:
+            sign_kernel = True
+
+    if sign_kernel:
+        linux_signed = tempfile.NamedTemporaryFile(prefix='linux-signed')
+        linux = linux_signed.name
+
+        cmd = [
+            *sbsign_invocation,
+            args.linux,
+            '--output', linux,
+        ]
+
+        print('+', shell_join(cmd))
+        subprocess.check_call(cmd)
+    else:
+        linux = args.linux
+
+    uki = UKI(args.stub)
+
+    # TODO: derive public key from from args.pcr_private_key?
+    pcrpkey = args.pcrpkey or args.pcr_public_key
+
+    sections = [
+        # name,     content,         measure?
+        ('osrel',   args.os_release, True ),
+        ('cmdline', args.cmdline,    True ),
+        ('dtb',     args.devicetree, True ),
+        ('splash',  args.splash,     True ),
+        ('pcrpkey', pcrpkey,         True ),
+        ('initrd',  args.initrd,     True ),
```

Comment:
Right now we use the file multiple times: possibly twice for measurement and once as input for `objcopy`. We could save it to a tempfile if the input is a pipe. We could also accept multiple `--initrd` params and concatenate them. We could also do both… I think is is more of an optional feature, so I'm leaving it for later. I'll add it to the TODO in comment.

Reply:
@medhefgo Maybe a bit too hacky, but `@<(cat initrd1 initrd2) should parse as the correct input, since it will passed in as the string `@/dev/fd/12345` and `/dev/fd/12345` will then end up in `args.initrd`. This might still break in other places, I only checked the initial argparse bit.